python/lsst/daf/butler/datastore/file_templates.py

```diff
@@ -449,7 +449,7 @@ def format(self, ref: DatasetRef) -> str:
         # context should only happen in unit tests; in general we need all
         # dimensions to fill out templates).
         fields: dict[str, object] = {
-            k: ref.dataId.get(k) for k in ref.datasetType.dimensions.names if ref.dataId.get(k) is not None
+            k: v for k in ref.datasetType.dimensions.names if (v := ref.dataId.get(k)) is not None
```

Comment:
```
fields = dict(dataId.mapping)
```
is all you need now.

[DM-21840](https://jira.lsstcorp.org/browse/DM-21840) is still open because `None` values in data IDs are not prohibited by DB constraints yet, but it's long been true that conditions like "exposure with NULL filter" is for all intents and purposes forbidden, because if you do it the exposure becomes unqueryable.